docs/RFCS/20200507_partial_indexes.md

```diff
@@ -0,0 +1,318 @@
+CREATE TABLE products (id INT PRIMARY KEY, price INT, units_sold INT, review_count INT)
+CREATE INDEX popular_prds ON t (price) WHERE units_sold > 1000 OR review_count > 100
+```
+
+No constraint can be created for the top-level predicate expression of
+`popular_prds`.
+
+Therefore, constraints alone cannot help us determine that `popular_prds` can be
+scanned to satisfy any of the below queries:
+
+```sql
+SELECT COUNT(id) FROM products WHERE units_sold > 1500 AND price > 100
+
+SELECT COUNT(id) FROM products WHERE review_count > 200 AND price < 100
+
+SELECT COUNT(id) FROM products WHERE (units_sold > 1000 OR review_count > 200) AND price < 100
+```
+
+In order to accommodate for such expressions, we must walk the filter and
+expression trees. At each predicate expression node, we will check if it is
+implied by the filter expression node.
+
+Postgres's [predtest library](https://github.com/postgres/postgres/blob/c9d29775195922136c09cc980bb1b7091bf3d859/src/backend/optimizer/util/predtest.c#L251-L287)
+uses this method to determine if a partial index can be used to satisfy a query.
+The logic Postgres uses for testing implication of conjunctions, disjunctions,
+and "atoms" (anything that is not an `AND` or `OR`) is as follows:
+
+    ("=>" means "implies")
+
+    atom A => atom B if:          A contains B
+    atom A => AND-expr B if:      A => each of B's children
+    atom A => OR-expr B if:       A => any of B's children
+
+    AND-expr A => atom B if:      any of A's children => B
+    AND-expr A => AND-expr B if:  A => each of B's children
+    AND-expr A => OR-expr B if:   A => any of B's children OR
+                                    any of A's children => B
+
+    OR-expr A => atom B if:       each of A's children => B
+    OR-expr A => AND-expr B if:   A => each of B's children
+    OR-expr A => OR-expr B if:    each of A's children => any of B's children
+
+The time complexity of this check is `O(P * F)`, where `P` is the number of
+nodes in the predicate expression and `F` is the number of nodes in the filter
+expression.
+
+## Generating Partial Index Scans
+
+We will consider utilizing partial indexes for both unconstrained and
+constrained scans. Therefore, we'll need to modify both the `GenerateIndexScans`
+and `GenerateConstrainedScans` exploration rules (or make new, similar rules).
+
+In addition, we'll need to update exploration rules for zig-zag joins and
+inverted index scans.
+
+We'll remove redundant filters from the expression when generating a scan over a
+partial index. For example:
+
+```sql
+CREATE TABLE products (id INT PRIMARY KEY, price INT, units_sold INT, units_in_stock INT)
+CREATE INDEX idx1 ON products (price) WHERE units_sold > 1000
+
+SELECT * FROM products WHERE price > 20 AND units_sold > 1000 AND units_in_stock > 0
+```
+
+When generating the constrained scan over `idx1`, the `units_sold > 1000` filter
+can be removed from the outer `Select`, such that only the `units_in_stock > 0`
+filter remains.
```

Comment:
I'd clarify that it can only be removed when the filter exactly matches, and also give example showing why we need to retain the filter when it doesn't exactly match (but where partial index can still be used), such as:
```
AND units_sold > 1200
```

Reply:
Done.